### PR TITLE
[BUGFIX] Prevent exception when tag already exists in the active or another asset collection

### PR DESCRIPTION
**Pull request by gerhard-boden**
Resolves: NEOS-1694

### CODE REVIEW

File: TYPO3.Media/Classes/TYPO3/Media/Controller/AssetController.php [hunk 1]
```diff
@@ -397,10 +397,6 @@ public function createTagAction($label)
         $existingTag = $this->tagRepository->findByLabel($label);
```
Comment (bwaidelich) [inline TYPO3.Media/Classes/TYPO3/Media/Controller/AssetController.php]: The issue is that `findBy..` returns a `QueryResult` (i.e. a *collection* of tags). It should be changed to
```
$existingTag = $this->tagRepository->findOneByLabel($label);
if ($existingTag !== null) {
 // ...
```

File: TYPO3.Media/Classes/TYPO3/Media/Controller/AssetController.php [hunk 2]
```diff
@@ -397,10 +397,6 @@ public function createTagAction($label)
...
         if (count($existingTag) > 0) {
             $this->addFlashMessage(sprintf('Tag "%s" already exists.', $label), '', \TYPO3\Flow\Error\Message::SEVERITY_ERROR);
-            if (($assetCollection = $this->browserState->get('activeAssetCollection')) !== null && $assetCollection->addTag($existingTag)) {
-                $this->assetCollectionRepository->update($assetCollection);
-                $this->addFlashMessage(sprintf('Tag "%s" added to collection.', $label));
-            }
```
Comment (bwaidelich) [inline TYPO3.Media/Classes/TYPO3/Media/Controller/AssetController.php]: The idea of this code was that an existing tag would be *added* to the current collection. Just deleting the code will eliminate this feature

**Comment by bwaidelich** [comment]
Thanks for taking care, but IMO this needs some adjustments (see inline comments)

File: TYPO3.Media/Classes/TYPO3/Media/Controller/AssetController.php (hunk 2 above)
Comment (gerhard-boden) [inline TYPO3.Media/Classes/TYPO3/Media/Controller/AssetController.php]: i see, did not see that. How will this that play out in the overview for all asset collections? Will there be duplicate tags? Could get confusing fast

File: TYPO3.Media/Classes/TYPO3/Media/Controller/AssetController.php (hunk 1 above)
Comment (gerhard-boden) [inline TYPO3.Media/Classes/TYPO3/Media/Controller/AssetController.php]: `$assetCollection->addTag($existingTag)` will only accept `TYPO3\Media\Domain\Model\Tag` so passing the query result caused the exception in the first place

File: TYPO3.Media/Classes/TYPO3/Media/Controller/AssetController.php (hunk 2 above)
Comment (bwaidelich) [inline TYPO3.Media/Classes/TYPO3/Media/Controller/AssetController.php]: just try it ;)
the tags won't be duplicated, but a tag can be part of multiple asset collections (m:m)

**Comment by gerhard-boden** [comment]
new commit: forget all my comments, after some testing I understand everything now! :) Thanks for your feedback! I removed the error message, since i think this could irritate the user, because actually nothing went wrong.

**Comment by gerhard-boden** [comment]
another thing:
In https://github.com/neos/neos-development-collection/blob/master/TYPO3.Media/Resources/Private/Templates/Asset/EditTag.html#l37 we should change the message to `<p>This will delete the tag across all asset collections, but not the assets that have it.<br />This operation cannot be undone.</p>`. I have no idea how to handle this in combination with the translation platform.

**Comment by bwaidelich** [comment]
:+1: by reading. thanks! Not sure about the error message you mentioned, but this is not exactly related anyways, is it?

**Comment by gerhard-boden** [comment]
it's not a real dealbreaker, but since the user can use the same tag across multiple asset collections with this fix, he maybe does not know that the tag deleted within his current asset collection will eventually remove the same tag in other asset collections

File: TYPO3.Media/Classes/TYPO3/Media/Controller/AssetController.php [hunk 3]
```diff
@@ -394,12 +394,11 @@ public function deleteAction(Asset $asset)
...
     {
-        $existingTag = $this->tagRepository->findByLabel($label);
-        if (count($existingTag) > 0) {
-            $this->addFlashMessage(sprintf('Tag "%s" already exists.', $label), '', \TYPO3\Flow\Error\Message::SEVERITY_ERROR);
+        $existingTag = $this->tagRepository->findOneByLabel($label);
+        if ($existingTag !== null) {
             if (($assetCollection = $this->browserState->get('activeAssetCollection')) !== null && $assetCollection->addTag($existingTag)) {
                 $this->assetCollectionRepository->update($assetCollection);
-                $this->addFlashMessage(sprintf('Tag "%s" added to collection.', $label));
+                $this->addFlashMessage(sprintf('Tag "%s" already exists and was added to collection.', $label));
```
Comment (kdambekalns) [inline TYPO3.Media/Classes/TYPO3/Media/Controller/AssetController.php]: This **MUST** be `htmlspecialchars($label)` to avoid another XSS vector.

**Comment by kdambekalns** [comment]
:-1: (see inline comment)

**Comment by gerhard-boden** [comment]
thanks for the advice @kdambekalns Did it for all flashmessages within this file, since all collection and tag names are user generated input

**Comment by kdambekalns** [comment]
Hi, thanks, awesome!

**Comment by aertmann** [comment]
:+1:

**Comment by daniellienert** [comment]
:+1:

**Comment by aertmann** [comment]
:+1: a bit of a mess with the commits, but oh well

**Pull request merged**